core/src/block_sync.rs

```diff
@@ -138,6 +163,8 @@ pub mod message {
         pub latest_hash: Option<HashOf<SignedBlock>>,
         /// Hash of second to latest block
         pub prev_hash: Option<HashOf<SignedBlock>>,
+        /// The block hashes already seen
+        pub seen_blocks: BTreeSet<HashOf<SignedBlock>>,
```

Comment:
I'm  a bit concerned that now `GetBlocksAfter` can grow potentially large but can't think of other good solution which would achieve the same thing.

One option could be to used randomized data structure like `LRUCache` to have bounded size for `GetBlocksAfter` and possibility of duplicated message send.

Reply:
malicious peer can send `GetBlocksAfter` with a huge list of invalid `seen_blocks`

Reply:
A malicious peer can send large amounts of invalid data in most of our network messages. So solving this kind of issue generally should be a separate ticket.

Reply:
then open a ticket for this

Reply:
Here is the ticket. https://github.com/hyperledger/iroha/issues/4944